File: Week Assignment 3 using Python/8_reverse_linked_list.py
# 8. Reverse linked list (iterative)
class Node:
    def __init__(self,data):
        self.data=data; self.next=None

def reverseList(head):
    prev = None; curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

if __name__ == "__main__":
    head = Node(1); head.next = Node(2); head.next.next = Node(3)
    head = reverseList(head)
    t = head
    while t:
        print(t.data, end=" ")
        t = t.next
    print()

File: Week Assignment 3 using C/8_reverse_linked_list.c
// 8. Reverse linked list (iterative)
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; };

struct Node* reverseList(struct Node* head) {
    struct Node *prev = NULL, *curr = head, *nxt = NULL;
    while (curr) {
        nxt = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nxt;
    }
    return prev;
}

void printList(struct Node* head) { while (head) { printf("%d ", head->data); head = head->next; } printf("\n"); }

int main() {
    struct Node* head = (struct Node*)malloc(sizeof(struct Node)); head->data=1; head->next=NULL;
    head->next = (struct Node*)malloc(sizeof(struct Node)); head->next->data=2; head->next->next=NULL;
    head->next->next = (struct Node*)malloc(sizeof(struct Node)); head->next->next->data=3; head->next->next->next=NULL;
    head = reverseList(head); // expected 3 2 1
    printList(head);
    return 0;
}
